iceoryx_utils/include/iceoryx_utils/internal/cxx/static_storage.inl

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_CXX_STATIC_STORAGE_INL
+#define IOX_UTILS_CXX_STATIC_STORAGE_INL
+
+#include "iceoryx_utils/internal/cxx/static_storage.hpp"
+
+namespace iox
+{
+namespace cxx
+{
+template <uint64_t Capacity, uint64_t Align>
+constexpr uint64_t static_storage<Capacity, Align>::align_delta(uint64_t align, uint64_t alignTarget) noexcept
+{
+    auto r = align % alignTarget;
+
+    // if r != 0 we are not aligned and need to add this amount to an align
+    // aligned address to be aligned with alignTarget
+    return r != 0 ? alignTarget - r : 0;
+}
+
+template <uint64_t Capacity, uint64_t Align>
+static_storage<Capacity, Align>::~static_storage() noexcept
+{
+    deallocate();
+}
+
+template <uint64_t Capacity, uint64_t Align>
+template <typename T>
+constexpr bool static_storage<Capacity, Align>::fits_statically() noexcept
+{
+    return sizeof(T) + align_delta(alignof(m_bytes), alignof(T)) <= Capacity;
+}
+
+template <uint64_t Capacity, uint64_t Align>
+template <typename T>
+T* static_storage<Capacity, Align>::allocate() noexcept
+{
+    static_assert(fits_statically<T>(), "type does not fit into static storage");
+    return reinterpret_cast<T*>(allocate(alignof(T), sizeof(T)));
+}
+
+template <uint64_t Capacity, uint64_t Align>
+void* static_storage<Capacity, Align>::allocate(uint64_t align, uint64_t size) noexcept
+{
+    if (m_ptr)
+    {
+        return nullptr; // cannot allocate, already in use
```

Comment:
i would opt against this, this comment is just for developers and it is common to return `nullptr` if allocation fails (or throw an exception which we do not want).  I would not output warnings for things like this. The way we use it, this should not happen since the storage is cleared when we want to store a different callable.